### DISCUSSION TITLE
Manually add a TradeLineItem to the InvoiceDescriptor?

### DISCUSSION THREAD

**Discussion by s1nn3r** [stephanstapel/ZUGFeRD-csharp#308] [General]
Hi Stephan,
I try to use your library with our ERP, which supports .net in a rather limited way.
Most functions a working fine. But I have a problem with using InvoiceDescriptor.AddTradeLineItem, as I there is no "null" in our ERP and I need to pass all parameters to the function or the ERP won't find the matching function in the library. So I cannot set things like billingPeriodStart to null.
So I thought about creating the TradeLineItem myself, just set the needed Parameters and then "link" it to the InvoiceDescriptor.
But I have no idea how I could do the "linking"? It there a way to do it?
Thanks!
Marc

**Comment by s1nn3r** [comment]
Sorry, found it myself:
InvoiceDescriptor.TradeLineItems.Add(TradeLineItem)

**Comment by s1nn3r** [comment]
I need to reopen this discussion, as I got stuck at applying a manual LineID to the TradeInvoiceLine via AssociatedDocument.
As it's "internal", I cannot create it and so I cannot apply it.

Wouldn't it be more flexible if the AssociatedDocument would be like:
        [...]
        public string LineID { get; set; }
        public AssociatedDocument(string lineID)
?
Thanks
Marc

**Comment by stephanstapel** [comment]
It might be easier for you just to create a small wrapper assembly that has exactly the functions you need (and you are able to call from your ERP) which then translates these calls into calls of the s2industries ZUGFeRD assembly. What do you think?

**Comment by s1nn3r** [comment]
Yes, I thought about that but after the manual creation of the TradeItemLine worked nicely I thought I could save myself that.
If the AssociatedDocument must stay internal for some causes, what about another function "AddTradeItemLine(String? LineID)" in InvoiceDescriptor that just creates an empty new line with optional LineID and returns it? After that one can fill the missing fields by hand. This would also be very flexible for systems that don't offer nullable dot net types.

**Comment by stephanstapel** [comment]
The whole library heavily relies on optional parameters as the ZUGFeRD/ XRechnung standards come with so many optional elements to cover (imho) edge cases. I'd rather not add more functions as a replacement to the mentioned optional parameters. I guess it would be much faster to create a simple and small wrapper library.
